README.md
CPR (cp -R)
===========

There are other modules out there that attempt this, but none did it the way I needed it to be done or they had issues and the author failed to fix them in a timely manner.

Install
-------

    npm install cpr


Build Status
------------

[![Build Status](https://secure.travis-ci.org/davglass/cpr.png)](http://travis-ci.org/davglass/cpr)

Usage
-----

```js
var cpr = require('cpr');
//or
var cpr = require('cpr').cpr; //Back compat

cpr('/path/from', '/path/to', {
    deleteFirst: true, //Delete "to" before
    overwrite: true, //If the file exists, overwrite it
    confirm: true //After the copy, stat all the copied files to make sure they are there
}, function(err, files) {
    //err - The error if any (err.list might be available with an array of errors for more detailed information)
    //files - List of files that we copied
});

cpr('/path/from', '/path/to', function(err, files) {
    //err - The error if any (err.list might be available with an array of errors for more detailed information)
    //      In the case of an error, cpr continues to copy files but returns this error object with all of the files that it failed to copy.
    //files - List of files that we copied
});
```

Options
-------

All options default to `false`.

   * `deleteFirst`: Delete the to directory with `rimraf`
   * `overwrite`: If the destination exists, overwrite it
   * `confirm`: After the copy operation, stat all the files and report errors if any are missing
   * `filter`: `RegExp` or `function` to test each file against before copying


Filtering
---------

If you give it a `RegExp`, it will use that to `test` the full absolute pathnames of the files and directories as they are being gathered. If any of them passes, it will not be copied.
If you give it a `function`, it will use that with `Array.filter` on the list of full absolute pathnames of files and directories.


CLI
---

`cpr` can also be used from the command line which is useful for cross platform support.

Usage:

```
cpr <source> <destination> [options]
```
Copies files from `source` to `destination`.

Options:

* `-d`, `--delete-first`:             Delete the destination directory before copying.
* `-f <regex>`, `--filter <regex>`:   Filter out any items that match `<regex>`, a case-insensitive regex pattern.
* `-h`, `--help`:                     Display this usage info.
* `-o`, `--overwrite`:                Overwrite the destination exists if it exists.
* `-v`, `--version`:                  Display the cpr version.

Note that the CLI's `filter` option is simpler than the original NodeJS API, only accepting case-insensitive regular expression patterns and not functions.

![cpr](../master/cpr.jpg?raw=true)


CHANGELOG.md
## Change Log

### v3.0.1 (2017/10/31 15:12 +00:00)
- [1c802c9](https://github.com/davglass/cpr/commit/1c802c942f85e244de266adec7c27524b0a0cb34) 3.0.1 (@davglass)
- [8d285b8](https://github.com/davglass/cpr/commit/8d285b89744e3bf3e9fb492c46a3879d5814d790) stablizing test (@davglass)
- [55fa41d](https://github.com/davglass/cpr/commit/55fa41ddfaaf0b86721b86a7a6b1524a9033acc2) adding Dockerfile for testing (@davglass)
- [008532b](https://github.com/davglass/cpr/commit/008532bb88edcf7477fa2b936d72cfe64fceca39) Dropping Node 4.x testing (@davglass)
- [66108f7](https://github.com/davglass/cpr/commit/66108f72b1729e046a2ac06c4898c5ee9ae2f967) changelog (@davglass)

### v3.0.0 (2017/10/31 13:08 +00:00)
- [e3cd84b](https://github.com/davglass/cpr/commit/e3cd84b7f27a61e8c86cd14379905781e0970806) 3.0.0 (@davglass)
- [17b0494](https://github.com/davglass/cpr/commit/17b0494af1b173a30e82e0d8c69297f4376d6cbb) Manually porting #43 (@davglass)
- [#57](https://github.com/davglass/cpr/pull/57) Fixing a race condition that can occur when a file has been read but the write buffer has not been flushed when copying a file. (#57) (@PlutoMatt)
- [#52](https://github.com/davglass/cpr/pull/52) Amend filtering info for more clarity. (#52) (@jrpool)
- [b0e8f40](https://github.com/davglass/cpr/commit/b0e8f40b303d9b4fbbd2f20e2853d0617ae48b19) added node.js specific build info (@davglass)
- [01efad1](https://github.com/davglass/cpr/commit/01efad136c554008fc529ff4f120fb863dfd1b8a) added node.js specific build info (@davglass)
- [e2395ba](https://github.com/davglass/cpr/commit/e2395ba3846fea420e02ed043cde014a52cd614a) added appveyor build config (@davglass)
- [#50](https://github.com/davglass/cpr/pull/50) various changes (#50) (@stevenvachon)

### v2.2.0 (2017/06/26 12:15 +00:00)
- [4520872](https://github.com/davglass/cpr/commit/4520872c30bbed3bac9930ad633f4dd4e5545225) 2.2.0 (@davglass)
- [0686492](https://github.com/davglass/cpr/commit/0686492dab8f1563172996dc1e1be5e39d0adac7) changelog (@davglass)
- [#48](https://github.com/davglass/cpr/pull/48) Various changes (#48) (@stevenvachon)
- [724900c](https://github.com/davglass/cpr/commit/724900c9ae5307cba2ba250fba5d1dfb74cb9af3) changelog (@davglass)

### v2.1.0 (2017/05/29 12:56 +00:00)
- [57e7c2e](https://github.com/davglass/cpr/commit/57e7c2ecd17842e490003ab4eabe07368de24ad5) 2.1.0 (@davglass)
- [8d3c003](https://github.com/davglass/cpr/commit/8d3c003e4658cb8c71d0f40167f96b0cc9f23f70) [fixes #44] - don't error when no callback given (@davglass)
- [#40](https://github.com/davglass/cpr/pull/40) Merge pull request #40 from oferh/cpr-typings (@oferh)
- [55dc201](https://github.com/davglass/cpr/commit/55dc2013cd00ae9ee5d97ae34d6a2f60382a306d) typings for cpr (@oferh)
- [#37](https://github.com/davglass/cpr/pull/37) Merge pull request #37 from davglass/tests (@davglass)
- [b4f8c00](https://github.com/davglass/cpr/commit/b4f8c003f2bd7f92e69e40c766846966ac34c484) fixing travis build (@davglass)
- [0f2c8b5](https://github.com/davglass/cpr/commit/0f2c8b51a33610d53f7f9604e55bea8fe6360966) changelog (@davglass)

### v2.0.2 (2016/12/21 18:14 +00:00)
- [3fa3af0](https://github.com/davglass/cpr/commit/3fa3af02572fe5a42fdf35ed879e3040fd209409) 2.0.2 (@davglass)
- [#36](https://github.com/davglass/cpr/pull/36) Merge pull request #36 from benelliott/master (@benelliott)
- [0443d51](https://github.com/davglass/cpr/commit/0443d511e92ee19a7d389e19d8f684f0ccb60c32) Include CLI documentation in README.md, update usage.txt to reflect CLI filter option (@benelliott)
- [14233bf](https://github.com/davglass/cpr/commit/14233bf74563aa90f7c1b5fe8094717aededaf72) Added changelog [skip ci] (@davglass)

### v2.0.1 (2016/12/19 15:47 +00:00)
- [4225fc6](https://github.com/davglass/cpr/commit/4225fc6ce0a62b8301ee881b9b2ff677246e8096) 2.0.1 (@davglass)
- [f05f162](https://github.com/davglass/cpr/commit/f05f16298eb7467eb3b741c1f62af95e3519f779) added contrib (@davglass)
- [#35](https://github.com/davglass/cpr/pull/35) Merge pull request #35 from benelliott/master (@benelliott)
- [42e6e57](https://github.com/davglass/cpr/commit/42e6e573ebecbb950dd033ad96c7790d6a798962) Add simplified version of `filter` functionality to CLI (@benelliott)
- [6ad0023](https://github.com/davglass/cpr/commit/6ad00236c6c5a888ea012c825d1422afa39e51fd) [Fixes #31] Added changelog [skip ci] (@davglass)

### v2.0.0 (2016/08/15 19:08 +00:00)
- [#30](https://github.com/davglass/cpr/pull/30) Merge pull request #30 from stevenvachon/master (@stevenvachon)
- [e7c4db1](https://github.com/davglass/cpr/commit/e7c4db18e36a10abb00d7eacdff08bced8fa9312) 2.0.0 (@stevenvachon)
- [82dcffb](https://github.com/davglass/cpr/commit/82dcffb40a5841c663a5e66e76876470ae8a3518) added cli options (@stevenvachon)
- [bf5d1ce](https://github.com/davglass/cpr/commit/bf5d1cec0a12c2a1841de58d501d29bcb45fbd58) adding new collaborator to package.json (@davglass)

### v1.1.2 (2016/06/09 14:26 +00:00)
- [0584e7c](https://github.com/davglass/cpr/commit/0584e7cf7135d13b79a1fbb8ccd9c68ef4614262) 1.1.2 (@davglass)
- [#29](https://github.com/davglass/cpr/pull/29) Merge pull request #29 from happinov/single-file-copy (@happinov)
- [a79261d](https://github.com/davglass/cpr/commit/a79261d39d1e832d76aa9f7d4e00b372b02f6314) Fixed behavior in case to has trailing path separator. (@ldanet)
- [128570a](https://github.com/davglass/cpr/commit/128570adfe3e3d2efd1de1b58147c4fb71bc32a3) Single files are now copied without creating a directory first. (@ldanet)

### v1.1.1 (2016/04/06 13:51 +00:00)
- [e691cb3](https://github.com/davglass/cpr/commit/e691cb365ca7db3113a5ef3487ad12df1a4088ea) 1.1.1 (@davglass)
- [e4bddfd](https://github.com/davglass/cpr/commit/e4bddfdad43d1d89ab4b7adbb43916411ef07c91) updating tests again (@davglass)
- [bf0f79a](https://github.com/davglass/cpr/commit/bf0f79a8c80f67c071107e09dbddaaa83c5221ab) fix tests based on npm 3.x file tree (@davglass)

tests/full.js
var assert = require('assert'),
    path = require('path'),
    fs = require('fs'),
    rimraf = require('rimraf'),
    mkdirp = require('mkdirp'),
    cpr = require('../lib'),
    exec = require('child_process').exec,
    to = path.join(__dirname, './out/'),
    from = path.join(__dirname, '../node_modules');

describe('cpr test suite', function() {
    this.timeout(55000);

    describe('loading', function() {
        before(function() {
            rimraf.sync(to);
        });

        it('should export raw method', function () {
            assert.equal(typeof cpr, 'function');
        });

        it('should export cpr method too', function () {
            assert.equal(typeof cpr.cpr, 'function');
        });
    });

    describe('should copy node_modules', function() {
        var out = path.join(to, '0');
        var data = {};
        var beforeStat = fs.statSync(path.join(from, '.bin/mocha'));

        before(function(done) {
            cpr(from, out, function(err, status) {
                data = {
                    from: fs.readdirSync(from).sort(),
                    to: fs.readdirSync(out).sort()
                };
                done();
            });
        });

        it('has ./out/0', function() {
            var stat = fs.statSync(out);
            assert.ok(stat.isDirectory());
        });

        it('dirs are equal', function() {
            assert.deepEqual(data.to, data.from);
        });

        it('from directory has graceful-fs dir', function() {
            var fromHasGFS = data.from.some(function(item) {
                return (item === 'graceful-fs');
            });
            assert.equal(true, fromHasGFS);
        });

        it('to directory has graceful-fs dir', function() {
            var toHasGFS = data.to.some(function(item) {
                return (item === 'graceful-fs');
            });
            assert.equal(true, toHasGFS);
        });

        it('preserves file mode of copied files', function () {
            var stat = fs.statSync(path.join(out, '.bin/mocha'));
            assert.equal(beforeStat.mode, stat.mode);
        });

    });

    describe('should NOT copy node_modules', function() {
        var out = path.join(to, '1'),
            data;

        before(function(done) {
            cpr(from, out, {
                filter: /node_modules/
            }, function(err) {
                fs.stat(out, function(e, stat) {
                    data = {
                        err: err,
                        stat: e
                    };
                    done();
                });
            });
        });

        it('does not have ./out/1', function() {
            assert.ok(data.stat); // Should be an error
        });
        it('threw an error', function() {
            assert(data.err instanceof Error); // Should be an error
            assert.equal(data.err.message, 'No files to copy');
        });

    });

    describe('should not copy yui-lint from regex', function() {
        var out = path.join(to, '2'),
            data;

        before(function(done) {
            cpr(from, out, {
                confirm: true,
                overwrite: true,
                filter: /yui-lint/
            }, function(err, status) {
                data = {
                    status: status,
                    dirs: {
                        from: fs.readdirSync(from).sort(),
                        to: fs.readdirSync(out).sort()
                    }
                };
                done();
            });
        });

        it('returns files array with confirm', function() {
            assert.ok(Array.isArray(data.status));
            assert.ok(data.status.length > 0);
        });
        it('has ./out/2', function() {
            var stat = fs.statSync(out);
            assert.ok(stat.isDirectory());
        });
        it('dirs are not equal', function() {
            assert.notDeepEqual(data.dirs.to, data.dirs.from);
        });
        it('from directory has yui-lint dir', function() {
            var fromHasLint = data.dirs.from.some(function(item) {
                return (item === 'yui-lint');
            });
            assert.equal(true, fromHasLint);
        });
        it('to directory does not have yui-lint dir', function() {
            var toHasLint = data.dirs.to.some(function(item) {
                return (item === 'yui-lint');
            });
            assert.equal(false, toHasLint);
        });
    });

    describe('should not copy directory from function', function() {
        var out = path.join(to, '3'),
            data;

        before(function(done) {
            cpr(from, out, {
                confirm: true,
                deleteFirst: true,
                filter: function (item) {
                    return !(/data/.test(item));
                }
            }, function(err, status) {
                data = {
                    status: status,
                    dirs: {
                        from: fs.readdirSync(path.join(from, 'jshint/')).sort(),
                        to: fs.readdirSync(path.join(out, 'jshint/')).sort()
                    }
                };
                done();
            });
        });

        it('has ./out/3', function() {
            var stat = fs.statSync(out);
            assert.ok(stat.isDirectory());
        });
        it('dirs are not equal', function() {
            assert.notDeepEqual(data.dirs.to, data.dirs.from);
        });
        it('from directory has data dir', function() {
            var fromHas = data.dirs.from.some(function(item) {
                return (item === 'data');
            });
            assert.equal(true, fromHas);
        });
        it('to directory does not have data dir', function() {
            var toHas = data.dirs.to.some(function(item) {
                return (item === 'data');
            });
            assert.equal(false, toHas);
        });

    });

    describe('should copy shelljs from bad filter', function() {
        var out = path.join(to, '4'),
            data;

        before(function(done) {
            cpr(from, out, {
                confirm: true,
                deleteFirst: true,
                filter: 'bs content'
            }, function(err, status) {
                data = {
                    status: status,
                    dirs: {
                        from: fs.readdirSync(path.join(from, 'jshint/node_modules')).sort(),
                        to: fs.readdirSync(path.join(out, 'jshint/node_modules')).sort()
                    }
                };
                done();
            });
        });
        it('has ./out/4', function() {
            var stat = fs.statSync(out);
            assert.ok(stat.isDirectory());
        });
        it('dirs are not equal', function() {
            assert.deepEqual(data.dirs.to, data.dirs.from);
        });
        it('from directory has shelljs dir', function() {
            var fromHas = data.dirs.from.some(function(item) {
                return (item === 'shelljs');
            });
            assert.equal(true, fromHas);
        });
        it('to directory does have shelljs dir', function() {
            var toHas = data.dirs.to.some(function(item) {
                return (item === 'shelljs');
            });
            assert.equal(true, toHas);
        });

    });

    describe('should copy node_modules with overwrite flag', function() {
        var out = path.join(to, '4'),
            data;

        before(function(done) {
            cpr(from, out, function() {
                cpr(from, out, {
                    overwrite: true,
                    confirm: true
                }, function(err, status) {
                    data = {
                        status: status,
                        dirs: {
                            from: fs.readdirSync(from).sort(),
                            to: fs.readdirSync(out).sort()
                        }
                    };
                    done();
                });
            });
        });

        it('should return files array', function() {
            assert.ok(Array.isArray(data.status));
            assert.ok(data.status.length > 0);
        });
        it('has ./out/0', function() {
            var stat = fs.statSync(out);
            assert.ok(stat.isDirectory());
        });
        it('dirs are equal', function() {
            assert.deepEqual(data.dirs.to, data.dirs.from);
        });
        it('from directory has graceful-fs dir', function() {
            var fromHasGFS = data.dirs.from.some(function(item) {
                return (item === 'graceful-fs');
            });
            assert.equal(true, fromHasGFS);
        });
        it('to directory has graceful-fs dir', function() {
            var toHasGFS = data.dirs.to.some(function(item) {
                return (item === 'graceful-fs');
            });
            assert.equal(true, toHasGFS);
        });

    });

    describe('error handling', function() {

        it('should fail on non-existant from dir', function(done) {
            cpr('./does/not/exist', path.join(to, 'does/not/matter'), function(err, status) {
                assert.equal(undefined, status);
                assert(err instanceof Error);
                assert.equal('From should be a file or directory', err.message);
                done();
            });
        });

        it('should fail on non-file', function(done) {
            cpr('/dev/null', path.join(to, 'does/not/matter'), function(err, status) {
                assert.equal(undefined, status);
                assert(err instanceof Error);
                assert.equal('From should be a file or directory', err.message);
                done();
            });
        });

        it('should return an error if a directory is to write over an existing file with the same name', function(done) {
            mkdirp.sync(path.join(to, 'empty-src2', 'a'));
            mkdirp.sync(path.join(to, 'empty-dest2'));
            fs.writeFileSync(path.join(to, 'empty-dest2', 'a'), 'FILE');
            cpr(path.join(to, 'empty-src2'), path.join(to, 'empty-dest2'), { overwrite: true }, function(errs) {
                var stat = fs.statSync(path.join(to, 'empty-dest2'));
                assert.ok(stat.isDirectory());
                assert.ok(errs);
                assert.ok(errs.list);
                assert.ok(errs.list[0]);
                assert.ok(errs.list[0].message.match(/exists and is not a directory, can not create/));
                assert.equal(errs.list[0].code, 'ENOTDIR');
                assert.equal(errs.list[0].errno, 27);
                done();
            });
        });


        it('should fail without write permissions', function(done) {
            /*istanbul ignore next - This is for docker perms*/
            (function() {
                if (process.getuid && process.getuid() === 0) {
                    console.log('Skipping this test as root has all access..');
                    return done();
                }
                var baddir = path.join(to, 'readonly');
                mkdirp.sync(baddir);
                fs.chmodSync(baddir, '555');
                cpr(path.join(from, '.bin'), baddir, function(errs, status) {
                    assert.ok(errs);
                    assert.ok(errs.list);
                    assert.ok(errs.list[0]);
                    assert.ok(errs.message.match(/Unable to copy directory entirely/));
                    done();
                });
            })();
        });

    });

    describe('validations', function() {

        it('should copy empty directory', function(done) {
            mkdirp.sync(path.join(to, 'empty-src'));
            cpr(path.join(to, 'empty-src'), path.join(to, 'empty-dest'), function() {
                var stat = fs.statSync(path.join(to, 'empty-dest'));
                assert.ok(stat.isDirectory());
                done();
            });
        });

        it('should not delete existing folders in out dir', function(done) {
            mkdirp.sync(path.join(to, 'empty-src', 'a'));
            mkdirp.sync(path.join(to, 'empty-dest', 'b'));
            cpr(path.join(to, 'empty-src'), path.join(to, 'empty-dest'), { overwrite: true }, function() {
                var stat = fs.statSync(path.join(to, 'empty-dest'));
                assert.ok(stat.isDirectory());
                var dirs = fs.readdirSync(path.join(to, 'empty-dest'));
                assert.equal(dirs[0], 'a');
                assert.equal(dirs[1], 'b');
                done();
            });
        });

        it('should copy one file', function(done) {
            cpr(__filename, path.join(to, 'one-file-test.js'), { overwrite: true }, function(err) {
                assert.equal(undefined, err);
                var stat = fs.statSync(path.join(to, 'one-file-test.js'));
                assert.ok(stat.isFile());
                done();
            });
        });

        it('should copy one file in dir if to has trailing sep', function(done) {
            cpr(__filename, path.join(to, 'one-file-dir'+path.sep), { overwrite: true }, function(err) {
                assert.equal(undefined, err);
                var stat = fs.statSync(path.join(to, 'one-file-dir','full.js'));
                assert.ok(stat.isFile());
                done();
            });
        });

        it('should not copy because file exists', function(done) {
            cpr(__filename, path.join(to, 'one-file-test.js'), function(err, status) {
                assert.equal(undefined, status);
                assert(err instanceof Error);
                assert.ok(/^File .* exists$/.test(err.message));
                assert.equal(err.code, 'EEXIST');
                assert.equal(err.errno, 47);
                done();
            });
        });

    });

    describe('should work as a standalone bin', function() {
        var out = path.join(to, '4'),
            data;

        before(function(done) {
            exec('node ./bin/cpr ' + from + ' ' + out, function(err) {
              data = {
                  dirs: {
                      from: fs.readdirSync(from).sort(),
                      to: fs.readdirSync(out).sort()
                  }
              };
              done();
            });
        });

        it('has ./out/4', function() {
            var stat = fs.statSync(out);
            assert.ok(stat.isDirectory());
        });
        it('dirs are equal', function() {
            assert.deepEqual(data.dirs.to, data.dirs.from);
        });
        it('from directory has graceful-fs dir', function() {
            var fromHasGFS = data.dirs.from.some(function(item) {
                return (item === 'graceful-fs');
            });
            assert.equal(true, fromHasGFS);
        });
        it('to directory has graceful-fs dir', function() {
            var toHasGFS = data.dirs.to.some(function(item) {
                return (item === 'graceful-fs');
            });
            assert.equal(true, toHasGFS);
        });

    });

});


lib/sync.js
/*
Copyright (c) 2012, Yahoo! Inc. All rights reserved.
Code licensed under the BSD License:
http://yuilibrary.com/license/
*/
var fs = require('graceful-fs');
var path = require('path');
var mkdirp = require('mkdirp').sync;
var cp = require('cp').sync;

var sync;

// copy a symlink
function symlinkSync(src, dest) {
    var link = fs.readlinkSync(src);
    fs.symlinkSync(link, dest);
}

// copy the contents of a directory
function dirSync(src, dest) {
    mkdirp(dest);
    fs.readdirSync(src).forEach(function (name) {
        sync(path.join(src, name), path.join(dest, name));
    });
}

// synchronous implementation
sync = module.exports = function (src, dest) {
    var stat = fs.lstatSync(src);

    // just copy files
    if (stat.isFile()) {
        return cp(src, dest);
    }

    // read the sym link and create a new one
    if (stat.isSymbolicLink()) {
        return symlinkSync(src, dest);
    }

    // note a file or a symlink?  probably a directory,
    // so we'll copy its contents
    dirSync(src, dest);
};


lib/index.js
/*
Copyright (c) 2012, Yahoo! Inc. All rights reserved.
Code licensed under the BSD License:
http://yuilibrary.com/license/
*/
var fs = require('graceful-fs');
var Stack = require('./stack').Stack;
var path = require('path');
var rimraf = require('rimraf');
var mkdirp = require('mkdirp');

var getTree = function(from, options, callback) {
    var stack = new Stack(),
        errors = [],
        results = {};

    options.stats = options.stats || {};
    options.toHash = options.toHash || {};

    fs.readdir(from, stack.add(function(err, dirs) {
        if (!dirs.length) {
            results[from] = true;
            fs.stat(from, stack.add(function(err, stat) {
                /*istanbul ignore next*/
                if (err) {
                    return errors.push(err);
                }
                options.stats[from] = stat;
                options.toHash[from] = path.join(options.to, path.relative(options.from, from));
            }));
        }
        dirs.forEach(function (dir) {
            var base = path.join(from, dir);
            fs.stat(base, stack.add(function(err, stat) {
                options.stats[base] = stat;
                options.toHash[base] = path.join(options.to, path.relative(options.from, base));
                /*istanbul ignore next*/
                if (err) {
                    return errors.push(err);
                }
                if (stat.isDirectory()) {
                    getTree(base, options, stack.add(function(errs, tree) {
                        /*istanbul ignore next*/
                        if (errs && errs.length) {
                            errs.forEach(function(item) {
                                errors.push(item);
                            });
                        }
                        //tree is always an Array
                        tree.forEach(function(item) {
                            results[item] = true;
                        });
                    }));
                } else {
                    results[base] = true;
                }
            }));
        });
    }));

    stack.done(function() {
        callback(errors, Object.keys(results).sort());
    });
};

var filterTree = function (tree, options, callback) {
    var t = tree;
    if (options.filter) {
        if (typeof options.filter === 'function') {
            t = tree.filter(options.filter);
        } else if (options.filter instanceof RegExp) {
            t = tree.filter(function(item) {
                return !options.filter.test(item);
            });
        }
    }
    callback(null, t);
};

var splitTree = function (tree, options, callback) {
    var files = {},
        dirs = {};

    tree.forEach(function(item) {
        var to = options.toHash[item];
        if (options.stats[item] && options.stats[item].isDirectory()) {
            dirs[item] = true;
        } else {
            dirs[path.dirname(item)] = true;
            options.stats[path.dirname(item)] = fs.statSync(path.dirname(item));
            options.toHash[path.dirname(item)] = path.dirname(to);
        }
    });

    tree.forEach(function(item) {
        if (!dirs[item]) {
            files[item] = true;
        }
    });

    callback(Object.keys(dirs).sort(), Object.keys(files).sort());
};


var createDirs = function(dirs, to, options, callback) {
    var stack = new Stack();

    dirs.forEach(function(dir) {
        var stat = options.stats[dir],
            to = options.toHash[dir];

        /*istanbul ignore else*/
        if (to && typeof to === 'string') {
            fs.stat(to, stack.add(function(err, s) {
                if (s && !s.isDirectory()) {
                    /*istanbul ignore next*/
                    err = new Error(to + ' exists and is not a directory, can not create');
                    /*istanbul ignore next*/
                    err.code = 'ENOTDIR';
                    /*istanbul ignore next*/
                    err.errno = 27;
                    options.errors.push(err);
                } else {
                    mkdirp(to, stat.mode, stack.add(function(err) {
                        /*istanbul ignore next*/
                        if (err) {
                            options.errors.push(err);
                        }
                    }));
                }
            }));
        }
    });

    stack.done(function() {
        callback();
    });
};

var copyFile = function(from, to, options, callback) {
    var dir = path.dirname(to);
    mkdirp(dir, function() {
        fs.stat(to, function(statError) {
            var err;
            if(!statError && options.overwrite !== true) {
                /*istanbul ignore next*/
                err = new Error('File '+to+' exists');
                /*istanbul ignore next*/
                err.code = 'EEXIST';
                /*istanbul ignore next*/
                err.errno = 47;
                return callback(err);
            }

            var fromFile = fs.createReadStream(from),
                toFile = fs.createWriteStream(to, {
                    mode: options.stats[from].mode
                }),
                called = false,
                cb = function(e) {
                    /*istanbul ignore next - This catches a hard to trap race condition */
                    if (!called) {
                        callback(e);
                        called = true;
                    }
                },
                /*istanbul ignore next*/
                onError = function (e) {
                    err = e;
                    cb(e);
                };

            fromFile.on('error', onError);
            toFile.on('error', onError);
            toFile.once('finish', function() {
                cb(err);
            });
            fromFile.pipe(toFile);
        });
    });
};

var createFiles = function(files, to, options, callback) {
    var next = process.nextTick,
        complete = 0,
        count = files.length,
        check = function() {
            /*istanbul ignore else - Shouldn't need this if graceful-fs does it's job*/
            if (count === complete) {
                callback();
            }
        },
        copy = function() {
            var from = files.pop(),
                to = options.toHash[from],
                bail;
            if (!from) {
                return check();
            }
            copyFile(from, to, options, function(err) {
                /*istanbul ignore next*/
                //This shouldn't happen with graceful-fs, but just in case
                if (/EMFILE/.test(err)) {
                    bail = true;
                    files.push(from);
                } else if (err) {
                    options.errors.push(err);
                }
                /*istanbul ignore next*/
                if (!bail) {
                    complete++;
                }
                next(copy);
            });
        };

    copy();
};

var confirm = function(files, options, callback) {
    var stack = new Stack(),
        errors = [],
        f = [],
        filtered = files;

    if (options.filter) {
        if (typeof options.filter === 'function') {
            filtered = files.filter(options.filter);
        } else if (options.filter instanceof RegExp) {
            filtered = files.filter(function(item) {
                return !options.filter.test(item);
            });
        }
    }

    /*istanbul ignore else - filtered should be an array, but just in case*/
    if (filtered.length) {
        filtered.forEach(function(file) {
            fs.stat(file, stack.add(function(err, stat) {
                /*istanbul ignore next*/
                if (err) {
                    errors.push(err);
                } else {
                    if (stat && (stat.isFile() || stat.isDirectory())) {
                        f.push(file);
                    }
                }
            }));
        });
    }

    stack.done(function() {
        /*istanbul ignore next */
        callback(((errors.length) ? errors : null), f.sort());
    });
};

var cpr = function(from, to, opts, callback) {
    if (typeof opts === 'function') {
        callback = opts;
        opts = {};
    }

    var options = {},
        proc;

    /*istanbul ignore next - in case a callback isn't provided*/
    callback = callback || function () {};

    Object.keys(opts).forEach(function(key) {
        options[key] = opts[key];
    });

    options.from = from;
    options.to = to;
    options.errors = [];

    proc = function() {
        getTree(options.from, options, function(err, tree) {
            filterTree(tree, options, function(err, t) {
                splitTree(t, options, function(dirs, files) {
                    if (!dirs.length && !files.length) {
                        return callback(new Error('No files to copy'));
                    }
                    createDirs(dirs, to, options, function() {
                        createFiles(files, to, options, function() {
                            var out = [], err;
                            Object.keys(options.toHash).forEach(function(k) {
                                out.push(options.toHash[k]);
                            });
                            if (options.confirm) {
                                confirm(out, options, callback);
                            } else if (!options.errors.length) {
                                callback(null, out.sort());
                            } else {
                                /*istanbul ignore next*/
                                err = new Error('Unable to copy directory' + (out.length ? ' entirely' : ''));
                                /*istanbul ignore next*/
                                err.list = options.errors;
                                /*istanbul ignore next*/
                                callback(err, out.sort());
                            }
                        });
                    });
                });
            });
        });
    };

    fs.stat(options.from, function(err, stat) {
        if (err) {
            return callback(new Error('From should be a file or directory'));
        }
        if (stat && stat.isDirectory()) {
            if (options.deleteFirst) {
                rimraf(to, function() {
                    proc();
                });
            } else {
                proc();

            }
        } else {
            if (stat.isFile()) {
                var dirRegex = new RegExp(path.sep + '$');
                if (dirRegex.test(to)) { // Create directory if has trailing separator
                  to = path.join(to, path.basename(options.from));
                }
                // ensure copyFile() can access cached stat
                options.stats = options.stats || {};
                options.stats[from] = stat;
                return copyFile(options.from, to, options, callback);
            }
            callback(new Error('From should be a file or directory'));
        }
    });
};

//Preserve backward compatibility
cpr.cpr = cpr;
//Export a function
module.exports = cpr;
